File: scala/healthcare/DeidentificationExample.scala
import AssertionDLApproachExample.spark
import com.johnsnowlabs.nlp.DocumentAssembler
import com.johnsnowlabs.nlp.annotator.{NerCrfModel, NerDLModel, SentenceDetector, WordEmbeddingsModel}
import com.johnsnowlabs.nlp.annotators.Tokenizer
import com.johnsnowlabs.nlp.annotators.ner.NerConverter
import com.johnsnowlabs.nlp.jsl.annotator.DeIdentification
import com.johnsnowlabs.nlp.training.CoNLL
import org.apache.spark.ml.Pipeline
import org.apache.spark.sql.SparkSession

object DeidentificationExample extends App {

  val spark: SparkSession = SparkSession
    .builder()
    .appName("test")
    .master("local[*]")
    .config("spark.driver.memory", "6G")
    .config("spark.kryoserializer.buffer.max", "1G")
    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    .getOrCreate()


  import spark.implicits._



  implicit val session = spark

  val trainDataSet = CoNLL().readDataset(spark, "src/test/resources/de-identification/train_dataset_main_small.csv")
  var nerDlModel = NerDLModel.pretrained().setOutputCol("ner").setInputCols("sentence", "token", "glove")
  var nerCrfModel = NerCrfModel.pretrained().setOutputCol("ner").setInputCols("sentence", "token", "pos", "glove")
  val embeddingsFile = "src/test/resources/ner-corpus/embeddings.100d.test.txt"

  val emptyDataset = Seq(
    ""
  ).toDS.toDF("text")

  val documentAssembler = new DocumentAssembler()
    .setInputCol("text")
    .setOutputCol("document")

  val sentenceDetector = new SentenceDetector()
    .setInputCols(Array("document"))
    .setOutputCol("sentence")
    .setUseAbbreviations(true)

  val tokenizer = new Tokenizer()
    .setInputCols(Array("sentence"))
    .setOutputCol("token")


  val embeddings = WordEmbeddingsModel.pretrained("embeddings_clinical", "en", "clinical/models")
    .setInputCols(Array("sentence", "token"))
    .setOutputCol("embeddings")

  val clinical_sensitive_entities = NerDLModel.pretrained("ner_deid_synthetic", "en", "clinical/models")
    .setInputCols(Array("sentence", "token", "embeddings")).setOutputCol("ner")

  val nerConverter = new NerConverter()
    .setInputCols(Array("sentence", "token", "ner"))
    .setOutputCol("ner_chunk")

  val deIdentification = new DeIdentification()
    .setInputCols(Array("ner_chunk", "token", "sentence"))
    .setOutputCol("dei")
    .setConsistentObfuscation(true)
    .setMode("obfuscate")
    .setObfuscateRefSource("faker")


  val pipeline = new Pipeline()
    .setStages(Array(
      documentAssembler,
      sentenceDetector,
      tokenizer,
      embeddings,
      clinical_sensitive_entities,
      nerConverter,
      deIdentification
    )).fit(emptyDataset)

  val testDataset = Seq(
    "Record date : 2093-01-13 , David Hale , M.D . , Name : Hendrickson , Ora MR . # 7194334 Date : 01/13/93 PCP : " +
      "Oliveira , 25 years-old , Record date : 2079-11-09 . Cocke County Baptist Hospital . 0295 Keats Street"
  ).toDS.toDF("text")

  val deIdentificationDataFrame = pipeline.transform(testDataset)
  val dataframe = deIdentificationDataFrame.select("dei.result")
  dataframe.show(truncate = false)
  spark.stop
}

File: java/healthcare/DeidentificationExample.java
import com.johnsnowlabs.nlp.AnnotatorApproach;
import com.johnsnowlabs.nlp.DocumentAssembler;
import com.johnsnowlabs.nlp.annotators.Tokenizer;
import com.johnsnowlabs.nlp.annotators.deid.DeIdentification;
import com.johnsnowlabs.nlp.annotators.deid.DeIdentificationModel;
import com.johnsnowlabs.nlp.annotators.ner.MedicalNerModel;
import com.johnsnowlabs.nlp.annotators.ner.NerConverterInternal;
import com.johnsnowlabs.nlp.annotators.ner.dl.NerDLModel;
import com.johnsnowlabs.nlp.annotators.sbd.pragmatic.SentenceDetector;
import com.johnsnowlabs.nlp.embeddings.WordEmbeddingsModel;
import org.apache.spark.ml.Pipeline;
import org.apache.spark.ml.PipelineModel;
import org.apache.spark.ml.PipelineStage;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import scala.collection.JavaConverters;

import java.util.Arrays;
import java.util.LinkedList;

public class DeidentificationExample {



    public static void main(String args[]) {
        SparkSession spark = SparkSession
                .builder()
                .appName("PipelineExample")
                .config("spark.master", "local")
                .config("spark.driver.memory", "6G")
                .config("spark.kryoserializer.buffer.max", "1G")
                .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
                .getOrCreate();
        LinkedList<String> text = new LinkedList<String>();

        text.add("My name is Jesus and Live in Spain.");

        Dataset<Row> data = spark.createDataset(text, Encoders.STRING()).toDF("text");

        DocumentAssembler document = new DocumentAssembler();
        document.setInputCol("text");
        document.setOutputCol("document");

        SentenceDetector sentenceDetector = new SentenceDetector();
        sentenceDetector.setInputCols(new String[]{"document"});
        sentenceDetector.setOutputCol("sentence");

        Tokenizer tokenizer = new Tokenizer();
        tokenizer.setInputCols(new String[]{"sentence"});
        tokenizer.setOutputCol("token");

        WordEmbeddingsModel wordEmbeddings = WordEmbeddingsModel.pretrained("embeddings_clinical", "en", "clinical/models");
        wordEmbeddings.setInputCols(new String[]{"sentence", "token"});
        wordEmbeddings.setOutputCol("embeddings");

        MedicalNerModel nerDlModel = MedicalNerModel.pretrained("ner_deid_large", "en", "clinical/models");
        nerDlModel.setInputCols(new String[]{"sentence", "token","embeddings"});
        nerDlModel.setOutputCol("ner");

        NerConverterInternal ner_chunk = new NerConverterInternal();
        ner_chunk.setInputCols(new String[]{"sentence", "token","ner"});
        ner_chunk.setOutputCol("ner_chunk");
        ner_chunk.setPreservePosition(false);

        DeIdentification deIdentification = new DeIdentification();
        deIdentification.setInputCols(JavaConverters.asScalaBuffer(Arrays.asList("sentence", "token", "ner_chunk")));
        deIdentification.setOutputCol("dei");
        deIdentification.setConsistentObfuscation(true);
        deIdentification.setMode("obfuscate");
        deIdentification.setObfuscateRefSource("faker");


        Pipeline pipeline = new Pipeline();
        pipeline.setStages(new PipelineStage[] {document, sentenceDetector,tokenizer,wordEmbeddings,nerDlModel,ner_chunk,deIdentification});

        PipelineModel pipelineModel = pipeline.fit(data);
        Dataset<Row> outputDf = pipelineModel.transform(data);
        outputDf.select("dei.result").show();







    }





}
